# USED CARS FROM EBAY - GERMANY

In this project, I'll work with a dataset of used cars from eBay "Kleinanzeigen": a classifieds section of the German eBay website.

You can find the data set here: https://data.world/data-society/used-cars-data

The aim of this project is to clean the data and analyze the included used car listings.

**1)**

    1.1 importing the libraries 

    1.2 read the dataset into pandas



In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1") 
# I have to choose Latin-1 enconding because UTF-8 was genarating error  


**2) ** 
    
    get to know the dataframe



In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

*OBSERVATIONS*:

- by analyzing the outcome I can assume the columns vehicleType, gearbox, model, fuelType and notRepairedDamage have some NaN values. That is why I should investigate those columns better. 


- some of the text is written in German


- The column names use camelcase instead of snakecase

**3)** 

    convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [3]:
print(autos.columns) #get to know column names 

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [4]:
# rename columns
autos_col = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = autos_col
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


**4)** 
    
    basic data exploration to determine what other cleaning tasks need to be done:
    
   - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis;
   - Examples of numeric data stored as text which can be cleaned and converted;
   - Any columns that need more investigation;


In [5]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Based on the output above, we see the following:

 - we can get rid of the nr_of_pictures column as all the numbers there are zero. 
 - registration_year is odd: min = 1000, max = 9999 and std = 105. This needs more cleaning.
 - power_ps has minimum of zero. Need to look at.
 - registration_month has minimum of zero instead of 1, so need to clean this.
 - Count for vehicle_type is 44905. So some are NaN or null.
 - gearbox count is 47320. So some are NaN or null.
 - Some cars do not have power_ps.
 - odometer looks to be saved as text (because of km and comma separator), this is why it could not calculate means,...etc.
 - seller and offer_type values are mostly the same.
 - Many cars have price zero and the column is also stored as text (because of the "$" sign and the comma separator).

**4.1** I´ll start by converting price and odometer text values to numbers (float).

In [6]:
autos["price"] = autos["price"].str.replace(',','')
autos["price"] = autos["price"].str.replace('$','').astype(float)

autos['odometer'] = (autos['odometer']
                                .str.replace('km','')
                                .str.replace(',','')
                                .astype(float))

# rename odometer column to odometer_km
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)


autos[["price", "odometer_km"]].head() #check

price  odometer_km
0  5000.0     150000.0
1  8500.0     150000.0
2  8990.0      70000.0
3  4350.0      70000.0
4  1350.0     150000.0

**4.2** I´ll analyze odometer_km and price columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove:

In [7]:
#to see how many unique values
print(autos["price"].unique().shape, 
      autos["odometer_km"].unique().shape)

# to view min/max/median/mean etc
print(autos["price"].describe(), 
      autos["odometer_km"].describe())


(2357,) (13,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [8]:
# to check outliers
print(autos["price"].value_counts().sort_index(ascending= False).head(20),
    autos["odometer_km"].value_counts(),)


99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64 150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64


In [9]:
# to check outliers
autos['price'].value_counts().sort_index(ascending= True).head(10)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64

We can see that there are some cars that probably are given for free ( that's possible). But there are also some suspicious expensive car at 99999999 and another one at 10000000. Those two are most likely outliers.

In [10]:
#removing outliers
autos[autos["price"] > 350000] = np.NaN

#test
autos["price"].value_counts().sort_index(ascending= False).head(20)

350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
250000.0    1
220000.0    1
198000.0    1
197000.0    1
194000.0    1
190000.0    1
180000.0    1
175000.0    1
169999.0    1
169000.0    1
163991.0    1
163500.0    1
155000.0    1
151990.0    1
Name: price, dtype: int64

**4.3** convert the string values on "date_crawled" column to numerical representation to understand the distributions:

In [11]:
# to get to know the roll
autos['date_crawled'][0:5]

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object

In [12]:
# to slice only the dates - first 10 characters:
autos['date_crawled'].str[:10].head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

In [13]:
# rank by date in ascending order, include missing values and use percentages instead of counts:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index())

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03326
2016-03-09    0.03320
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03676
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03150
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03748
2016-03-22    0.03290
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03414
2016-03-30    0.03362
2016-03-31    0.03190
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03648
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
NaN           0.00028
Name: date_crawled, dtype: float64

**4.4** Dealing with Incorrect Registration Year Data: 

In [14]:
# understand the distribution of registration_year
autos["registration_year"].describe()


count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

- As we can see the minimum value is 1000 (before cars were invented) and maximum value is 9999 (many years into the future), this are odd year numbers of registration. 

- Also, because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

- Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely. 


In [17]:
autos = autos[autos["registration_year"].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True)

2000.0    0.069852
2005.0    0.062792
1999.0    0.062438
2004.0    0.057002
2003.0    0.056794
2006.0    0.056377
2001.0    0.056273
2002.0    0.052753
1998.0    0.051087
2007.0    0.047984
2008.0    0.046464
2009.0    0.043673
1997.0    0.042236
2011.0    0.034030
2010.0    0.033260
1996.0    0.030073
2012.0    0.027553
2016.0    0.027408
1995.0    0.027324
2013.0    0.016786
2014.0    0.013850
1994.0    0.013745
1993.0    0.009268
2015.0    0.008310
1990.0    0.008226
1992.0    0.008122
1991.0    0.007414
1989.0    0.003770
1988.0    0.002957
1985.0    0.002166
            ...   
1977.0    0.000458
1966.0    0.000458
1975.0    0.000396
1969.0    0.000396
1965.0    0.000354
1964.0    0.000250
1910.0    0.000187
1963.0    0.000187
1959.0    0.000146
1961.0    0.000125
1956.0    0.000104
1962.0    0.000083
1958.0    0.000083
1937.0    0.000083
1950.0    0.000062
1934.0    0.000042
1941.0    0.000042
1951.0    0.000042
1954.0    0.000042
1955.0    0.000042
1957.0    0.000042
1952.0    0.

We can see that most of the cars are from the period from 1994 to 2016. Another thing worth mentioned is the fact that there are 3 periods that represent main sets of offers:

    year: 2000 with 0.069842%
    year: 2005 with 0.062782%
    year: 1999 with 0.062449%

**5)** 

    Exploring Price by Brand
    
- explore variations across brands with aggregation technique;
- get the frequency and absolute number of each brand.

In [24]:
#  select the top 20:
autos["brand"].value_counts().head(20)


volkswagen        10185
bmw                5283
opel               5194
mercedes_benz      4579
audi               4149
ford               3350
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
smart               668
citroen             668
toyota              599
sonstige_autos      523
hyundai             473
volvo               444
mini                415
Name: brand, dtype: int64

In [28]:
# key for the dictionary of the top 20 cars
autos['brand'].value_counts().index[:20]

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')

**5.1** We'll loop over our selected brands, and assign the mean price to the dictionary, with the brand name as the key. By this, we will know the mean price for each brand from top 20

In [27]:
mean_price_for_top_20_cars = {}

for e in autos['brand'].value_counts().index[:20]: 
    top_brands = autos[autos["brand"] == e]
    mean_price = top_brands['price'].mean()
    mean_price_for_top_20_cars[e] = mean_price
    
for key in mean_price_for_top_20_cars:
    print(key, ':', mean_price_for_top_20_cars[key])

fiat : 2711.8011272141707
peugeot : 3039.4682651622
sonstige_autos : 10805.078393881453
toyota : 5115.33388981636
opel : 2876.716403542549
audi : 9093.65003615329
skoda : 6334.91948051948
nissan : 4664.891034482758
mazda : 4010.7716643741405
hyundai : 5308.53911205074
bmw : 8102.536248343744
smart : 3542.706586826347
volkswagen : 5231.081983308787
seat : 4296.492554410081
volvo : 4757.108108108108
mini : 10460.012048192772
citroen : 3699.935628742515
renault : 2395.4164467897976
ford : 3652.095223880597
mercedes_benz : 8485.239571958942


In [37]:
# print the same dictionary sorted 

sorted_dict_mean = (sorted(mean_price_for_top_20_cars
            .items(), key=lambda kv: kv[1], reverse = True))

sorted_dict_mean


[('sonstige_autos', 10805.078393881453),
 ('mini', 10460.012048192772),
 ('audi', 9093.65003615329),
 ('mercedes_benz', 8485.239571958942),
 ('bmw', 8102.536248343744),
 ('skoda', 6334.91948051948),
 ('hyundai', 5308.53911205074),
 ('volkswagen', 5231.081983308787),
 ('toyota', 5115.33388981636),
 ('volvo', 4757.108108108108),
 ('nissan', 4664.891034482758),
 ('seat', 4296.492554410081),
 ('mazda', 4010.7716643741405),
 ('citroen', 3699.935628742515),
 ('ford', 3652.095223880597),
 ('smart', 3542.706586826347),
 ('peugeot', 3039.4682651622),
 ('opel', 2876.716403542549),
 ('fiat', 2711.8011272141707),
 ('renault', 2395.4164467897976)]

We can observe in the top 7 brands there is a distinct price gap:

 - Mini, Audi, BMW and Mercedes Benz are more expensive
 - Renault and Fiat are less expensive
 - Volvo is in between

**5.2** understand the average mileage for those cars and if there's any visible link with mean price.

   - combine the data from both series objects (price and  odometer_km) into a single dataframe (with a shared index brand") and display the dataframe directly. 

In [38]:
mean_mileage_for_top_20_cars = {}

for e in autos['brand'].value_counts().index[:20]: 
    top_brands = autos[autos["brand"] == e]
    mean_mileage = top_brands['odometer_km'].mean()
    mean_mileage_for_top_20_cars[e] = mean_mileage
    
for key in mean_mileage_for_top_20_cars:
    print(key, ':', mean_mileage_for_top_20_cars[key])

fiat : 116553.94524959743
peugeot : 127136.81241184767
sonstige_autos : 87466.53919694072
toyota : 115709.51585976628
opel : 129223.14208702349
audi : 129287.78018799711
skoda : 110954.54545454546
nissan : 118572.41379310345
mazda : 124745.5295735901
hyundai : 106511.62790697675
bmw : 132431.38368351315
smart : 99595.80838323354
volkswagen : 128724.10407461954
seat : 121563.57388316152
volvo : 138355.85585585586
mini : 88602.40963855422
citroen : 119580.8383233533
renault : 128183.81706244503
ford : 124068.65671641791
mercedes_benz : 130856.0821139987


In [41]:
# Convert both dictionaries to series objects:
bpp = pd.Series(mean_price_for_top_20_cars)
bpm = pd.Series(mean_mileage_for_top_20_cars)
bpm #test

audi              129287.780188
bmw               132431.383684
citroen           119580.838323
fiat              116553.945250
ford              124068.656716
hyundai           106511.627907
mazda             124745.529574
mercedes_benz     130856.082114
mini               88602.409639
nissan            118572.413793
opel              129223.142087
peugeot           127136.812412
renault           128183.817062
seat              121563.573883
skoda             110954.545455
smart              99595.808383
sonstige_autos     87466.539197
toyota            115709.515860
volkswagen        128724.104075
volvo             138355.855856
dtype: float64

In [45]:
# Create a dataframe from the first series object using the dataframe constructor:
bpp_bpm_df = pd.DataFrame(bpp, columns=['mean_price'])
bpp_bpm_df

mean_price
audi             9093.650036
bmw              8102.536248
citroen          3699.935629
fiat             2711.801127
ford             3652.095224
hyundai          5308.539112
mazda            4010.771664
mercedes_benz    8485.239572
mini            10460.012048
nissan           4664.891034
opel             2876.716404
peugeot          3039.468265
renault          2395.416447
seat             4296.492554
skoda            6334.919481
smart            3542.706587
sonstige_autos  10805.078394
toyota           5115.333890
volkswagen       5231.081983
volvo            4757.108108

In [46]:
# Assign the other series as a new column in this dataframe:
bpp_bpm_df['mean_mileage'] = bpm
(bpp_bpm_df.sort_values(by = 'mean_price', 
                             ascending = False))

mean_price   mean_mileage
sonstige_autos  10805.078394   87466.539197
mini            10460.012048   88602.409639
audi             9093.650036  129287.780188
mercedes_benz    8485.239572  130856.082114
bmw              8102.536248  132431.383684
skoda            6334.919481  110954.545455
hyundai          5308.539112  106511.627907
volkswagen       5231.081983  128724.104075
toyota           5115.333890  115709.515860
volvo            4757.108108  138355.855856
nissan           4664.891034  118572.413793
seat             4296.492554  121563.573883
mazda            4010.771664  124745.529574
citroen          3699.935629  119580.838323
ford             3652.095224  124068.656716
smart            3542.706587   99595.808383
peugeot          3039.468265  127136.812412
opel             2876.716404  129223.142087
fiat             2711.801127  116553.945250
renault          2395.416447  128183.817062

**CONCLUSION**

There is no clear correlation between price and mileage. It looks the average price is more determined by the brand name if the difference in mileage is within 20,000 km. 